### ISSUE TITLE
Froyo A2SD

### ISSUE THREAD

**Issue by AnderWeb** [AnderWeb/android_packages_apps_Launcher#13]
Some users report that sometimes apps installed to SD card shows twice in the app drawer.
Seems some desktop icons/widgets from apps on SD are failing too...

**Comment by zmarty** [comment]
I have the same problem (the 2nd problem from above) with apps stored on the SD card.

Apps icons: All icons from apps stored on the SD card are missing from my home screen. They are replaced by the stock/default green alien icon. Restarting ADW fixes the problem.

Widgets: All widgets stored on the SD card fail to load. Restarting ADW launcher does NOT fix the problem.

**Comment by AnderWeb** [comment]
Please make sure you read AND UNDERSTAND this http://android-developers.blogspot.com/2010/07/apps-on-sd-card-details.html  before complaining on A2SD things.
Thanks :-)

**Comment by zmarty** [comment]
Hi there,

I fail to understand why this RTFM would help with my problem. The document states that ACTION_EXTERNAL_APPLICATIONS_AVAILABLE is broadcasted as an intent when a card is mounted. ADW Launcher might act on this or it might not, but the bottom line is that the icons are not loaded the first time, until I restart ADW Launcher. I am not the only one with this issue, see comments here: http://jbthemes.com/anderweb/. I would have commented there but the comments are closed.

What exactly did I miss in that document?

Thanks,
Ovi

**Comment by zmarty** [comment]
Just for clarification: the card is already in the phone when I start it up.

**Comment by zmarty** [comment]
Also, I am using the apps2sd from CM6, which I'm pretty sure is not the Google one.

**Comment by AnderWeb** [comment]
1-when you "are pretty sure of something" you should be really  pretty sure. Cm6 a2sd is the google a2sd.
2-RTFM again, especially the LAST poin.
3-On my github code there's already a fix for the apps showing/hiding in response to  the broadcasted intents.
4-if you think you can do it better, please do it and send patches, this is a free AND opensource project, I'm not apple, you cannot come here complaining like an angry kid.

**Comment by AnderWeb** [comment]
1-when you "are pretty sure of something" you should be really  pretty sure. Cm6 a2sd is the google a2sd.
2-RTFM again, especially the LAST poin.
3-On my github code there's already a fix for the apps showing/hiding in response to  the broadcasted intents.
4-if you think you can do it better, please do it and send patches, this is a free AND opensource project, I'm not apple, you cannot come here complaining like an angry kid.

**Comment by garok89** [comment]
correct me if i am wrong....but is it not rather clearly stated that widgets and apps which require background services should not be stored on the sdcard?
widgets fail to load when on the sdcard because the sdcard fat32 partition isnt mounted until after it is checked (which is after the launcher is started)
apps on sd fail to appear for the same reason...if memory serves, this wouldnt be an issue on stock froyo as it wouldnt check the fat32 partition before mounting it
this wasnt the case with app icons in the past because of the way apps2sd was implimented....

before complaining - try to gain a basic understanding of what you are complaining about

**Comment by garok89** [comment]
one more point....

IT IS NOT A GREEN ALIEN....IT IS AN ANDROID....IE. A ROBOT WHICH IS HUMANOID IN APPEARANCE

**Comment by zmarty** [comment]
Wow thanks for the hissy fit, you're making the community great. Look, I am also a programmer and I understand when you get frustrated, but what if I were a normal clueless user? This is not the way to treat your users, even if this is an open source project.

"but is it not rather clearly stated that widgets and apps which require background services should not be stored on the sdcard?" - where exactly is that stated in the document?

"before complaining - try to gain a basic understanding of what you are complaining about" - so what if I understand the issue? It's still a bug and it's valid.

**Comment by zmarty** [comment]
Also, why exactly do you have a way to submit bugs if you clearly don't welcome bug reports? This is ridiculous.

**Comment by garok89** [comment]
well.....

"When not to install on SD card?

The advantage of installing on SD card is easy to understand: contention for storage space is reduced. There are costs, the most obvious being that your app is disabled when the SD card is either removed or in USB Mass Storage mode; this includes running Services, not just interactive Activities. Aside from this, device removal disables an application’s Widgets, Input methods, Account Managers, Device administrators, Live wallpapers, and Live folders, and may require explicit user action to re-enable them."

the issue is not an issue with ADW...it is an issue with how apps2sd works and the fact that most custom roms do not mount the fat32 partition at the same point as stock android does

**Comment by AnderWeb** [comment]
As said in the last point of the link i gave you, "disables an application’s Widgets, Input methods, Account Managers, Device administrators, Live wallpapers, and Live folders, and may require explicit user action to re-enable them"

I don't have a problem with this being a bug, my problem is with your "angry complaining" manners. The issue is open, what means I (or whoever) should find a better solution. This is not a forum to add dozens of "mee too" posts. Its an issue tracker (oh, you as programmer should know.... ).
So if the issue is open and you're NOT providing useful info to resolve it, don't post here, use forums or whatever.

P.S.-Do you know howto properly submit a bug report? As you are programmer... you should know that too....

**Comment by zmarty** [comment]
Hi garok89. I know this. I understand that the card is fully loaded only after the launcher is. And I understand there is a patch that reloads the icons when the launcher gets the broadcast. This leaves one issue, why don't the widgets load correctly even after I restart ADW?

Thanks for your work.

**Comment by garok89** [comment]
widgets that fail to load have never been fixed for me when reloading ADW...or stock, for that matter (even when not on sd)
as far as i can figure, it is more to do with the way the widgets start, and not the way the launcher grabs them...but i could be wrong

**Comment by zmarty** [comment]
@Anderweb: Ok, I agree it was a me too post and I apologize if that is inappropriate. I just wanted to confirm the bug and add the part about the widgets. I still believe my message added useful information to the issue. But why did you send me to the manual? What good did it do? Again, thanks for all you do and I apologize.

@garok89: Thanks for the info.

**Comment by garok89** [comment]
i believe he sent you to manual because you spoke of having widgets on sd....
assuming that wasnt sarcas, you are welcome

**Comment by zmarty** [comment]
@garok89: That was not sarcasm. Regarding the last point from the manual you cited, it only says: "may require explicit user action to re-enable them.". It does not say that widgets do not load or run from sd cards.

**Comment by garok89** [comment]
using logic.....if the fat32 isnt mounted at boot time until after the launcher is loaded, the apps there are unavailable to for the launcher to load them

an earlier revision of that page (or a similar page, i am not sure) explicitly stated that they should not be put on the sdcard because of the issues

sorry if i seemed hostile, i am just so used to people on XDA who know nothing wanting to be spoon fed or complaining about stuff which is well known and easily found on google - they really irritate me now (i wasnt always like this....it wears you down) (it doesnt help that i also run the cmupdater support line and get the same degree of stupidity, if not worse)

**Comment by zmarty** [comment]
@garok89: In understand perfectly, thanks again.

**Comment by AnderWeb** [comment]
C'mon guys, leave the chat for forumsssss :P

**Comment by Macrophage001** [comment]
Figured out the double app problem. Just move it back to phone and then to sdcard or vise versa.

**Comment by elrubens** [comment]
Same problem here:

ADW.Launcher does not show the icons of the applications stored in SDCard (just normal shortcut icons, no widgets). As expected, if I remove the shortcut and create it again the icon shows correctly, but it is lost after every restart.

I've checked this with default home launcher from Froyo and the problem does not happen.

ADW Version: 1.0.1 Standalone
Nexus 1
2.2 (Froyo) (FRF91) (No cyanogen)

Cheers

**Comment by AnderWeb** [comment]
Partially fixed on latest commits. Desktop icons now properly updates when external storage is available, but i need to work on ActionButton's icons and DockBar icons too.

**Comment by AnderWeb** [comment]
Ok, fully fixed.
I close this, because i'm not sure if the launcher should take care of the widgets installed on SDcard (read previous comments of the issue).

Could someone test the stock froyo launcher and see what happen with it about widgets moved to SD?

**Comment by rogerchucker** [comment]
When you close an issue does the fix show up in the next release, i.e. 1.0.2? I'm running 1.0.1 and it seems to still have the issue. Thanks a bunch.

**Comment by elrubens** [comment]
Thanks!

regarding widgets.. none of the applications I can install on sdcard has widgets.. anyway if someone gives an example (app name) I'll be happy to test

**Comment by AnderWeb** [comment]
Fixes after a "version info" commit are usually released on the next release.
Widget developers should avoid granting user to move the app to sd, but some devs or old widgets doesn't do this, but I don't currently know a sample :(

**Comment by rogerchucker** [comment]
Thanks for the clarification. As an FYI, I just noticed that once I moved the Astrid app to the SD card, the Astrid widget which was earlier populated with tasks turned empty. And then removing that widget when I tried to reinstate the widget, I couldn't find Astrid from the Widget list anymore (the menu that appears on long-pressing the home). So this all happened even without a reboot. Am I correct in assuming that this is sort of related to the issue for which this thread was started? I'm running ADW 1.0.1. I was a little confused if Astrid belongs to the group of apps that Google recommends against moving to SD card (widgets, background services?). I am also wondering if this is an issue only because Astrid in general is not ready yet to be moved to SD card but CyanogenMod's universal option to move "everything" to SD card is overriding the Astrid developer's settings. Thanks for listening!
